tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,253 @@
+use sharded_slab::{Guard, Slab};
+
+use super::stack::SpanStack;
+use crate::{
+    registry::{
+        extensions::{Extensions, ExtensionsInner, ExtensionsMut},
+        LookupSpan, SpanData,
+    },
+    sync::RwLock,
+};
+use std::{
+    cell::RefCell,
+    sync::atomic::{fence, AtomicUsize, Ordering},
+};
+use tracing_core::{
+    dispatcher,
+    field::FieldSet,
+    span::{self, Current, Id},
+    Event, Interest, Metadata, Subscriber,
+};
+
+/// A shared, reusable store for spans.
+///
+/// This registry is implemented using a [lock-free sharded slab][1], and is
+/// highly optimized for concurrent access.
+///
+/// [1]: https://docs.rs/crate/sharded-slab/0.0.5
+#[derive(Debug)]
+pub struct Registry {
+    spans: Slab<DataInner>,
+}
+
+/// Span data stored in a [`Registry`].
+///
+/// This definition  is intentionally kept small—only data pertaining to span
+/// relationships, span metadata, and active references are stored. Additional
+/// data, such as formatted fields, may be stored in the [extensions] typemap.
```

Comment:
This doesn't really tell me _why_ some stuff should be stored in the registry instead of extensions. What's the heuristic I should use to determine where something should go?

Reply:
I _think_ the answer is that you only ever get `&` to things stored in this data, whereas you can get `&mut` to anything in an extension (though at the cost of taking a lock). That'd be good to explain here.

Reply:
The registry stores the well-known data defined by `tracing`, extensions are for user-defined data of arbitrary types.

Reply:
Then the docs should say that :)

Reply:
@davidbarsky what do you think of changing "This definition is intentionally kept small" (which I didn't ever really care for) with something like 

> The registry stores well-known data defined by `tracing`: span relationships, metadata and reference counts. Additional user-defined data provided by [`Layer`s], such as formatted fields, may be stored in the [extensions] typemap.